File: ios/Photo/UI/CropOverlayView.swift
import UIKit

/// Handle gestures resize the box; interior gestures pan/pinch the media, keeping the box fixed.
final class CropOverlayView: UIView {
  var onCropChanged: ((CGFloat, CGFloat, CGFloat, CGFloat) -> Void)?
  var onMediaBoundsChanged: ((CGRect) -> Void)?
  var onViewportChanged: ((CGFloat, CGFloat, CGFloat) -> Void)?
  private var imageBounds = CGRect.zero, fittedBounds = CGRect.zero, cropRect = CGRect.zero
  private var aspectRatio: CGFloat?
  private var activeHandle: CropGeometry.Handle?
  private var primary: UITouch?, secondary: UITouch?
  private var lastPoint = CGPoint.zero
  private var lastDistance: CGFloat = 0

  override init(frame: CGRect) {
    super.init(frame: frame); backgroundColor = .clear; isOpaque = false; isMultipleTouchEnabled = true
  }
  required init?(coder: NSCoder) { nil }

  func restore(bounds: CGRect, state: PhotoTransformState) {
    guard bounds.width > 0, bounds.height > 0 else { return }
    fittedBounds = bounds
    let zoom = max(1, min(8, state.zoom))
    let cx = bounds.midX + state.panX*bounds.width, cy = bounds.midY + state.panY*bounds.height
    imageBounds = CGRect(x: cx-bounds.width*zoom/2, y: cy-bounds.height*zoom/2, width: bounds.width*zoom, height: bounds.height*zoom)
    aspectRatio = state.aspectRatio
    setCrop(left: state.cropLeft, top: state.cropTop, right: state.cropRight, bottom: state.cropBottom)
    onMediaBoundsChanged?(imageBounds)
  }
  func setImageBounds(_ bounds: CGRect, resetCrop: Bool) {
    let old = normalized()
    imageBounds = bounds; fittedBounds = bounds
    if resetCrop || cropRect.isEmpty { cropRect = bounds }
    else { setCrop(left: old.minX, top: old.minY, right: old.maxX, bottom: old.maxY) }
    setNeedsDisplay()
  }
  func setCrop(left: CGFloat, top: CGFloat, right: CGFloat, bottom: CGFloat) {
    cropRect = CGRect(x: imageBounds.minX+left*imageBounds.width, y: imageBounds.minY+top*imageBounds.height,
      width: (right-left)*imageBounds.width, height: (bottom-top)*imageBounds.height)
    setNeedsDisplay()
  }
  func setAspectRatio(_ ratio: CGFloat?) {
    aspectRatio = ratio
    cropRect = CropGeometry.fit(imageBounds.intersection(bounds), ratio: ratio)
    reportCrop(); setNeedsDisplay()
  }

  override func draw(_ rect: CGRect) {
    guard !cropRect.isEmpty, let context = UIGraphicsGetCurrentContext() else { return }
    let shade = UIBezierPath(rect: bounds); shade.append(UIBezierPath(rect: cropRect)); shade.usesEvenOddFillRule = true
    UIColor.black.withAlphaComponent(0.65).setFill(); shade.fill()
    context.setStrokeColor(UIColor.white.cgColor); context.setLineWidth(1); context.stroke(cropRect)
    context.setStrokeColor(UIColor.white.withAlphaComponent(0.33).cgColor); context.setLineWidth(0.5)
    for i in 1...2 {
      let x = cropRect.minX+cropRect.width*CGFloat(i)/3, y = cropRect.minY+cropRect.height*CGFloat(i)/3
      context.move(to: CGPoint(x:x,y:cropRect.minY)); context.addLine(to:CGPoint(x:x,y:cropRect.maxY))
      context.move(to: CGPoint(x:cropRect.minX,y:y)); context.addLine(to:CGPoint(x:cropRect.maxX,y:y))
    }
    context.strokePath(); context.setStrokeColor(UIColor.white.cgColor); context.setLineWidth(3); context.setLineCap(.round)
    for x in [cropRect.minX,cropRect.maxX] {
      for y in [cropRect.minY,cropRect.maxY] {
        context.move(to: CGPoint(x:x+(x == cropRect.minX ? 14 : -14),y:y)); context.addLine(to:CGPoint(x:x,y:y))
        context.addLine(to: CGPoint(x:x,y:y+(y == cropRect.minY ? 14 : -14)))
      }
    }
    for y in [cropRect.minY,cropRect.maxY] { context.move(to:CGPoint(x:cropRect.midX-8,y:y)); context.addLine(to:CGPoint(x:cropRect.midX+8,y:y)) }
    for x in [cropRect.minX,cropRect.maxX] { context.move(to:CGPoint(x:x,y:cropRect.midY-8)); context.addLine(to:CGPoint(x:x,y:cropRect.midY+8)) }
    context.strokePath()
  }

  override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
    for touch in touches { if primary == nil { primary = touch } else if secondary == nil { secondary = touch } }
    guard let primary else { return }
    if secondary != nil { activeHandle = nil } else { activeHandle = hitHandle(primary.location(in:self)) }
    let metrics = touchMetrics(); lastPoint = metrics.0; lastDistance = metrics.1
  }
  override func touchesMoved(_ touches: Set<UITouch>, with event: UIEvent?) {
    guard primary != nil, imageBounds.width > 0 else { return }
    let (point,distance) = touchMetrics(), dx = point.x-lastPoint.x, dy = point.y-lastPoint.y
    if let activeHandle {
      cropRect = CropGeometry.resize(cropRect, bounds: imageBounds.intersection(bounds), handle: activeHandle, dx: dx, dy: dy, ratio: aspectRatio, minimum: 48)
    } else {
      moveMedia(dx:dx,dy:dy,factor:lastDistance > 0 && distance > 0 ? distance/lastDistance : 1,focus:point)
    }
    lastPoint = point; lastDistance = distance; reportCrop(); setNeedsDisplay()
  }
  override func touchesEnded(_ touches: Set<UITouch>, with event: UIEvent?) {
    if let secondary, touches.contains(secondary) { self.secondary = nil }
    if let primary, touches.contains(primary) { self.primary = secondary; secondary = nil }
    let metrics = touchMetrics(); lastPoint = metrics.0; lastDistance = metrics.1
    if primary == nil { activeHandle = nil }
  }
  override func touchesCancelled(_ touches: Set<UITouch>, with event: UIEvent?) { primary = nil; secondary = nil; activeHandle = nil }

  private func moveMedia(dx:CGFloat,dy:CGFloat,factor:CGFloat,focus:CGPoint) {
    let minimum = max(1,max(cropRect.width/fittedBounds.width,cropRect.height/fittedBounds.height))
    let zoom = imageBounds.width/fittedBounds.width
    let scale = max(minimum,min(max(8,minimum),zoom*factor))/zoom
    let w = imageBounds.width*scale, h = imageBounds.height*scale
    let x = max(cropRect.maxX-w,min(cropRect.minX,focus.x+(imageBounds.minX-focus.x)*scale+dx))
    let y = max(cropRect.maxY-h,min(cropRect.minY,focus.y+(imageBounds.minY-focus.y)*scale+dy))
    imageBounds = CGRect(x:x,y:y,width:w,height:h)
    onMediaBoundsChanged?(imageBounds)
    onViewportChanged?(w/fittedBounds.width,(imageBounds.midX-fittedBounds.midX)/fittedBounds.width,(imageBounds.midY-fittedBounds.midY)/fittedBounds.height)
  }
  private func touchMetrics() -> (CGPoint,CGFloat) {
    guard let a = primary?.location(in:self) else { return (.zero,0) }
    guard let b = secondary?.location(in:self) else { return (a,0) }
    return (CGPoint(x:(a.x+b.x)/2,y:(a.y+b.y)/2),hypot(a.x-b.x,a.y-b.y))
  }
  private func hitHandle(_ p:CGPoint) -> CropGeometry.Handle? {
    let points: [(CGPoint,CropGeometry.Handle)] = [
      (CGPoint(x:cropRect.minX,y:cropRect.minY),.topLeft),(CGPoint(x:cropRect.maxX,y:cropRect.minY),.topRight),
      (CGPoint(x:cropRect.minX,y:cropRect.maxY),.bottomLeft),(CGPoint(x:cropRect.maxX,y:cropRect.maxY),.bottomRight),
      (CGPoint(x:cropRect.midX,y:cropRect.minY),.top),(CGPoint(x:cropRect.maxX,y:cropRect.midY),.right),
      (CGPoint(x:cropRect.midX,y:cropRect.maxY),.bottom),(CGPoint(x:cropRect.minX,y:cropRect.midY),.left)]
    guard let nearest = points.min(by: { hypot(p.x-$0.0.x,p.y-$0.0.y) < hypot(p.x-$1.0.x,p.y-$1.0.y) }), hypot(p.x-nearest.0.x,p.y-nearest.0.y) <= 24 else { return nil }
    return nearest.1
  }
  private func normalized() -> CGRect {
    guard imageBounds.width > 0,imageBounds.height > 0 else { return CGRect(x:0,y:0,width:1,height:1) }
    return CGRect(x:(cropRect.minX-imageBounds.minX)/imageBounds.width,y:(cropRect.minY-imageBounds.minY)/imageBounds.height,width:cropRect.width/imageBounds.width,height:cropRect.height/imageBounds.height)
  }
  private func reportCrop() { let r = normalized(); onCropChanged?(max(0,r.minX),max(0,r.minY),min(1,r.maxX),min(1,r.maxY)) }
}

File: android/src/main/java/com/photovideoeditor/photo/ui/CropOverlayView.kt
package com.photovideoeditor.photo.ui

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.view.MotionEvent
import android.view.View
import com.photovideoeditor.photo.render.CropGeometry
import com.photovideoeditor.photo.render.PhotoTransformState
import kotlin.math.abs
import kotlin.math.hypot

/** Crop handles own resize gestures; interior drags/pinches move the media behind a fixed box. */
class CropOverlayView(context: Context) : View(context) {
  private var imageBounds = RectF()
  private var fittedBounds = RectF()
  private var cropRect = RectF()
  private var aspectRatio: Float? = null
  var onCropChanged: ((Float, Float, Float, Float) -> Unit)? = null
  var onMediaBoundsChanged: ((RectF) -> Unit)? = null
  var onViewportChanged: ((Float, Float, Float) -> Unit)? = null
  private val density = resources.displayMetrics.density
  private val dimPaint = Paint().apply { color = Color.argb(165, 0, 0, 0) }
  private val borderPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.WHITE; style = Paint.Style.STROKE; strokeWidth = density }
  private val gridPaint = Paint().apply { color = Color.argb(85, 255, 255, 255); strokeWidth = density * .5f }
  private val handlePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.WHITE; strokeWidth = 3 * density; strokeCap = Paint.Cap.ROUND; style = Paint.Style.STROKE }
  private var activeHandle: CropGeometry.Handle? = null
  private var lastX = 0f; private var lastY = 0f; private var lastDistance = 0f
  private var movingMedia = false

  fun restore(bounds: RectF, state: PhotoTransformState) {
    if (bounds.width() <= 0 || bounds.height() <= 0) return
    fittedBounds = RectF(bounds)
    val zoom = state.zoom.coerceIn(1f, 8f)
    val cx = bounds.centerX() + state.panX * bounds.width()
    val cy = bounds.centerY() + state.panY * bounds.height()
    imageBounds = RectF(cx - bounds.width() * zoom / 2, cy - bounds.height() * zoom / 2,
      cx + bounds.width() * zoom / 2, cy + bounds.height() * zoom / 2)
    aspectRatio = state.aspectRatio
    setCrop(state.cropLeft, state.cropTop, state.cropRight, state.cropBottom)
    onMediaBoundsChanged?.invoke(RectF(imageBounds))
    invalidate()
  }

  fun setImageBounds(bounds: RectF, resetCrop: Boolean) {
    val old = normalized()
    imageBounds = RectF(bounds); fittedBounds = RectF(bounds)
    if (resetCrop || cropRect.isEmpty) cropRect = RectF(bounds) else setCrop(old[0], old[1], old[2], old[3])
    invalidate()
  }

  fun setCrop(left: Float, top: Float, right: Float, bottom: Float) {
    cropRect = RectF(imageBounds.left + left * imageBounds.width(), imageBounds.top + top * imageBounds.height(),
      imageBounds.left + right * imageBounds.width(), imageBounds.top + bottom * imageBounds.height())
    invalidate()
  }

  fun setAspectRatio(ratio: Float?) {
    aspectRatio = ratio
    // A new preset starts with the largest centered rectangle in the visible media.
    val available = RectF(imageBounds)
    available.intersect(0f, 0f, width.toFloat(), height.toFloat())
    cropRect = CropGeometry.fit(available.geometry(), ratio).native()
    reportCrop(); invalidate()
  }

  override fun onDraw(canvas: Canvas) {
    if (cropRect.isEmpty) return
    canvas.drawRect(0f, 0f, width.toFloat(), cropRect.top, dimPaint)
    canvas.drawRect(0f, cropRect.bottom, width.toFloat(), height.toFloat(), dimPaint)
    canvas.drawRect(0f, cropRect.top, cropRect.left, cropRect.bottom, dimPaint)
    canvas.drawRect(cropRect.right, cropRect.top, width.toFloat(), cropRect.bottom, dimPaint)
    canvas.drawRect(cropRect, borderPaint)
    for (i in 1..2) {
      val x = cropRect.left + cropRect.width() * i / 3; val y = cropRect.top + cropRect.height() * i / 3
      canvas.drawLine(x, cropRect.top, x, cropRect.bottom, gridPaint)
      canvas.drawLine(cropRect.left, y, cropRect.right, y, gridPaint)
    }
    val corner = 14 * density
    for ((x, y) in listOf(cropRect.left to cropRect.top, cropRect.right to cropRect.top, cropRect.left to cropRect.bottom, cropRect.right to cropRect.bottom)) {
      canvas.drawLine(x, y, x + if (x == cropRect.left) corner else -corner, y, handlePaint)
      canvas.drawLine(x, y, x, y + if (y == cropRect.top) corner else -corner, handlePaint)
    }
    val edge = 8 * density
    for (y in listOf(cropRect.top, cropRect.bottom)) canvas.drawLine(cropRect.centerX() - edge, y, cropRect.centerX() + edge, y, handlePaint)
    for (x in listOf(cropRect.left, cropRect.right)) canvas.drawLine(x, cropRect.centerY() - edge, x, cropRect.centerY() + edge, handlePaint)
  }

  override fun onTouchEvent(event: MotionEvent): Boolean {
    if (imageBounds.isEmpty) return false
    when (event.actionMasked) {
      MotionEvent.ACTION_DOWN -> {
        activeHandle = hitTest(event.x, event.y); movingMedia = activeHandle == null
        lastX = event.x; lastY = event.y; lastDistance = 0f
        parent?.requestDisallowInterceptTouchEvent(true)
      }
      MotionEvent.ACTION_POINTER_DOWN -> {
        activeHandle = null; movingMedia = true
        lastX = (event.getX(0) + event.getX(1)) / 2; lastY = (event.getY(0) + event.getY(1)) / 2
        lastDistance = hypot(event.getX(0) - event.getX(1), event.getY(0) - event.getY(1))
      }
      MotionEvent.ACTION_MOVE -> {
        val multi = event.pointerCount > 1
        val x = if (multi) (event.getX(0) + event.getX(1)) / 2 else event.x
        val y = if (multi) (event.getY(0) + event.getY(1)) / 2 else event.y
        val distance = if (multi) hypot(event.getX(0) - event.getX(1), event.getY(0) - event.getY(1)) else 0f
        val handle = activeHandle
        if (handle != null) {
          val valid = RectF(imageBounds).apply { intersect(0f, 0f, width.toFloat(), height.toFloat()) }
          cropRect = CropGeometry.resize(cropRect.geometry(), valid.geometry(), handle, x - lastX, y - lastY, aspectRatio, 48 * density).native()
        } else if (movingMedia) {
          moveMedia(x - lastX, y - lastY, if (lastDistance > 0 && distance > 0) distance / lastDistance else 1f, x, y)
        }
        lastX = x; lastY = y; lastDistance = distance
        reportCrop(); invalidate()
      }
      MotionEvent.ACTION_POINTER_UP -> {
        val index = if (event.actionIndex == 0) 1 else 0
        lastX = event.getX(index); lastY = event.getY(index); lastDistance = 0f
      }
      MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
        activeHandle = null; movingMedia = false; lastDistance = 0f
        parent?.requestDisallowInterceptTouchEvent(false)
      }
    }
    return true
  }

  private fun moveMedia(dx: Float, dy: Float, factor: Float, focusX: Float, focusY: Float) {
    val minimumZoom = maxOf(1f, cropRect.width() / fittedBounds.width(), cropRect.height() / fittedBounds.height())
    val zoom = imageBounds.width() / fittedBounds.width()
    val scale = (zoom * factor).coerceIn(minimumZoom, maxOf(8f, minimumZoom)) / zoom
    val w = imageBounds.width() * scale; val h = imageBounds.height() * scale
    val left = (focusX + (imageBounds.left - focusX) * scale + dx).coerceIn(cropRect.right - w, cropRect.left)
    val top = (focusY + (imageBounds.top - focusY) * scale + dy).coerceIn(cropRect.bottom - h, cropRect.top)
    imageBounds = RectF(left, top, left + w, top + h)
    onMediaBoundsChanged?.invoke(RectF(imageBounds))
    onViewportChanged?.invoke(w / fittedBounds.width(), (imageBounds.centerX() - fittedBounds.centerX()) / fittedBounds.width(), (imageBounds.centerY() - fittedBounds.centerY()) / fittedBounds.height())
  }

  private fun hitTest(x: Float, y: Float): CropGeometry.Handle? {
    val points = listOf(
      Triple(cropRect.left, cropRect.top, CropGeometry.Handle.TOP_LEFT), Triple(cropRect.right, cropRect.top, CropGeometry.Handle.TOP_RIGHT),
      Triple(cropRect.left, cropRect.bottom, CropGeometry.Handle.BOTTOM_LEFT), Triple(cropRect.right, cropRect.bottom, CropGeometry.Handle.BOTTOM_RIGHT),
      Triple(cropRect.centerX(), cropRect.top, CropGeometry.Handle.TOP), Triple(cropRect.right, cropRect.centerY(), CropGeometry.Handle.RIGHT),
      Triple(cropRect.centerX(), cropRect.bottom, CropGeometry.Handle.BOTTOM), Triple(cropRect.left, cropRect.centerY(), CropGeometry.Handle.LEFT))
    val nearest = points.minByOrNull { hypot(x - it.first, y - it.second) } ?: return null
    return nearest.third.takeIf { hypot(x - nearest.first, y - nearest.second) <= 24 * density }
  }
  private fun normalized(): FloatArray = if (imageBounds.isEmpty) floatArrayOf(0f, 0f, 1f, 1f) else floatArrayOf(
    (cropRect.left - imageBounds.left) / imageBounds.width(), (cropRect.top - imageBounds.top) / imageBounds.height(),
    (cropRect.right - imageBounds.left) / imageBounds.width(), (cropRect.bottom - imageBounds.top) / imageBounds.height())
  private fun reportCrop() { val r = normalized(); onCropChanged?.invoke(r[0].coerceIn(0f, 1f), r[1].coerceIn(0f, 1f), r[2].coerceIn(0f, 1f), r[3].coerceIn(0f, 1f)) }
  private fun RectF.geometry() = CropGeometry.Rect(left, top, right, bottom)
  private fun CropGeometry.Rect.native() = RectF(left, top, right, bottom)
}
